### ISSUE TITLE
RFC: Improve sorting control for gridfield lists/pagination 

### ISSUE THREAD

**Issue by nicolewilliams16** [symbiote/silverstripe-gridfieldextensions#188]
The problem: 
When using Gridfield, items can not be ordered with drag and drop if they exist across different pages.
 
While it is possible to set the items per page higher, this can instead lead to having too many items on one page, and needs developer input. It’s a binary approach which creates further issues eg a long list or having to go back for developer changes as the list grows 

Who has this problem: 
Raised by 2 public sector users

Agency 1- can not order a list of 11 intranet content categories alphabetically on live site as the list spreads over two pages  
Agency 2 - attempting to order lists of documents for download 

Proposed solution: 
Give greater control to sort the list view to CMS users, by allowing the view to be adjusted different numbers of items per page. 
 
This could be following approach of Google Analytics as per screenshots or by adding a free text field for more flexibility of items shown. 

![screen shot 2017-05-17 at 12 40 52 pm](https://cloud.githubusercontent.com/assets/11512277/26134513/54e7752c-3b03-11e7-81f2-cf8dd2cdc71f.png)
![screen shot 2017-05-17 at 12 40 47 pm](https://cloud.githubusercontent.com/assets/11512277/26134514/54e9a022-3b03-11e7-87d3-33bd9b66cd0b.png)
![screen shot 2017-05-17 at 12 40 40 pm](https://cloud.githubusercontent.com/assets/11512277/26134512/54e76dc0-3b03-11e7-953c-9e7f5b62f062.png)

**Comment by wilr** [comment]
As long as developers have a way to specify the max number (i.e loading 5000 members on a single page might break servers) I'd love to see a component for it.

**Comment by nicolewilliams16** [comment]
Thanks for the feedback @wilr, that suggests that the Google version with item number options is a better approach than free text then :)

**Comment by michalkleiner** [comment]
Would it be more accurate saying that the problem is with 'moving items between pages in gridfields with more than one page'? That is the issue we were dealing with for our gov client.

Looking at it from that perspective, the solution could perhaps be to fix the mechanism for moving items across pages? Introducing a solution to dynamically adjust the number of items per page won't fix the broken functionality for moving items to the next or previous page.

As a workaround, we currently use https://github.com/ntd/silverstripe-togglepaginator to toggle the pagination, but that can obviously run into server resources issue with high numbers of items.

**Comment by michalkleiner** [comment]
As another workaround for the time being, we have a number field in site settings so we can adjust the number of records per page for one or more gridfields. It's not ideal, but worth mentioning.

**Comment by nicolewilliams16** [comment]
Thanks @michalkleiner, you're right the main pain point this fixes is the ability to move items across pages. 

I'm aware it's possible to adjust at the developer level but this is problematic currently because
1) users aren't aware that this could be changed by a developer OR
2) if they are, they often don't want to pay for this to be fixed 

Thanks for suggesting the alternative togglepaginator module. Toggling between 10 or every record could work for some applications, but its not a very flexible solution for the user, especially if there's a large number of items (as is the case for Agency 2).

**Comment by micmania1** [comment]
Some things to think about:
1. If the limit is not dev controlled, a user could potentially bring down their own website
2. server limits will probably prevent the above, but the request won't succeed and will throw an error.
3. server limits by themselves won't prevent it bring down your database (and therefore website)

My suspicion is that SilverStripe won't be able to handle that many items per page anyway so before implementation you should test to see how many items it can handle on production-like environments. (ie. max timeout = 30 seconds, memory correctly limited etc).

Another potential solution could be an infinite scroll with a "Start from no. xxx" option. That would prevent the above problems and allow for the user to see as much as they need on a single page.

**Comment by sminnee** [comment]
I think the control makes sense. A developer-defined dropdown of options seems better than a free text box for UX and practicality.

It might be easier to implement this as a complete replacement of the pagination component rather than something that interacts with it, but see how you get on.

I'd probably have the default dropdown options max out at 100/page.

**Comment by nyeholt** [comment]
On the subject of dragging across pages, `GridFieldOrderableRows` already supports dragging between pages; the drop-target could probably be made bigger / highlighted better, but the functionality is already there. 

![image](https://cloud.githubusercontent.com/assets/161730/26136178/ce209474-3afd-11e7-9700-b69bbe958428.png)

**Comment by michalkleiner** [comment]
@nyeholt yes, but it doesn't work correctly. Neither does the one from GridFieldSortableRows.

**Comment by michalkleiner** [comment]
I didn't do any further investigation as to what could be the actual cause (whether it's something around has_many or many_many_extraFields or similar), but, in general, items get reordered randomly even on pages of the list not involved in the current move. Also, it's not clear where the item should land on the target page - top or bottom.

**Comment by nyeholt** [comment]
@michalkleiner I think that's due to the sort orders of all items being set relative to each other in context of the list (and pages), which may not be correct in a global all-items-of-that-type context. Either way, it's still a problem.

**Comment by lerni** [comment]
scope creep: Sorting versioned objects or indicate staged Objects would also be of great value. A colored sort-handler similar to draft-lables in site-tree would probable help.

**Comment by oddnoc** [comment]
What about taking a step back from pagination and implement scrolling? That would improve the UX of browsing records and also solve the problem of reordering across arbitrary page boundaries.

**Comment by jonom** [comment]
I like the dropdown approach. If we're mainly talking about re-ordering functionality, I wonder if there is a practical limit to how many rows you can include before re-ordering becomes a pain? Like if you showed 1000 records, dragging something from the bottom of the page to the top might make for some long and awkward mouse work if you have to use the scroll wheel while holding down a button.

Say there is a practical limit of 100 rows per page and you have more than that, perhaps in that case, once you start dragging a row, the pagination bar could be replaced with a drop zone labelled 'move to page'. If you drop it there you get prompted to either choose a page to move the item to, or enter a position out of the total number of items. Then load that page afterwards so you can see the result and fine tune it.

From the sound of what @nyeholt said it this is probably similar to the behaviour in GridFieldOrderableRows does but more flexible and obvious/intuitive.

**Comment by robbieaverill** [comment]
A quick update/recap on the suggestions so far:

> @wilr: As long as developers have a way to specify the max number (i.e loading 5000 members on a single page might break servers) I'd love to see a component for it.

> @micmania1: If the limit is not dev controlled, a user could potentially bring down their own website
> My suspicion is that SilverStripe won't be able to handle that many items per page anyway so before implementation you should test to see how many items it can handle on production-like environments. (ie. max timeout = 30 seconds, memory correctly limited etc).

If a solution is implemented that allows anybody to produce large datasets, the approach should have some benchmarking done to ensure that the higher limit can feasibly be achieved without pushing near default server timeout limits (e.g. 30 seconds). This in itself may dictate what the upper limit we allow by default is.

Pages might be good for benchmarking because they have a fair number of attributes by default.

> @sminnee: I'd probably have the default dropdown options max out at 100/page.

So perhaps the default step sizes should be 10, 20, 50, 100. Personally I'd tend to agree that a 100-per-page maximum is generous enough as a default.

> @sminnee: A developer-defined dropdown of options seems better than a free text box for UX and practicality.

So the approach would be a dropdown containing pre-defined (configurable) step sizes, but no free-text input. 👍 

---

## Possible approach

We add a new GridField component that adds a new dropdown somewhere near the "View 1-30 of 90" in the footer of a GridField that says something like "Page size: [dropdown]".

The global default values of this dropdown will be dev configurable, e.g.:

```yaml
GridFieldPageSizes:
  step_sizes: [10, 20, 30, 50, 100]
```

You could also set custom step sizes on a per-instance basis at call time via public API:

```php
$gridField->getConfig()
    ->getComponentByType('GridFieldPageSizes')
    ->setStepSizes(array(50, 100, 200, 500));
```

The component will not require the `GridFieldPaginator` component to be in the config but would work with or without it.

---

How do we feel about an approach like this?

**Comment by michalkleiner** [comment]
What happens if the number of records displayed by default is not aligned with the dropdown options? Let's say the default GF has 15 rows and the dropdown only offers 10, 20, 50, how do I go back to the initial view without reloading the page?

Also, this new component is not fixing the problem of correctly moving items between pages, might be worth mentioning that.

Otherwise it looks really good and will be a great addition to the default set of GF components!

**Comment by robbieaverill** [comment]
I suppose in that context it would need to replace the default paginator. They obviously would work in tandem.

**Comment by robbieaverill** [comment]
>Also, this new component is not fixing the problem of correctly moving items between pages, might be worth mentioning that.
Good point.

**Comment by nicolewilliams16** [comment]
The core problem, is ordering items. Moving across pages is the current blocker but solving ordering can be solved in different ways, one of those is make it easier to move across pages.

**Comment by fspringveldt** [comment]
What about adding a line on each row titled _move to page:_? 
It might clutter up the UI in that it has a dropdown per row but using that in-conjunction with the _GridFieldPageSizes_ mentioned above it could do the trick.

**Comment by robbieaverill** [comment]
>What about adding a line on each row titled move to page:?

The existing drag and drop to previous and next page concept is possibly a better idea than doing this. It's possibly broken at the moment or only working in some scenarios, but this isn't part of this RFC.

**Comment by fspringveldt** [comment]
Drag + drop [GridFieldOrderableRows] also just sends the item to a page +1 or -1 of the current page which should be kept in mind.

**Comment by clarkepaul** [comment]
With a dropdown to adjust the number of viewable items, you won't be able to move anything from the first 100 to the second page of 100 items, or am I missing something?

**Comment by nicolewilliams16** [comment]
@clarkepaul depends on the dev configured options, but the ability to move between pages was seen as less of the problem than being able to reorder longer lists. We haven't yet seen a scenario where it would be feasible to order 100 items but allowing the developer to configure the options allows the use case and performance implications to be balanced.

**Comment by clarkepaul** [comment]
It would solve the main issue for sure, and a great first step. 

An option to improve the ability to drag between pages could be that when you grab an item to drag you provide two drop areas "prev page" and "next page" in place of the standard pagination (a simplified view than what was mentioned above).

**Comment by michalkleiner** [comment]
@clarkepaul that's exactly what both OrderableRows and SortableRows extensions do. They provide drop areas for both previous and next page. One in the footer, one on the sides. But neither does work in a 100% correct way, mostly around edge cases like the first or the last page. It's also not clear where it should go to within the destination page.

IMHO fixing that and making it predictable and easy to use (perhaps increasing the size of the drop areas or highlighting them better etc.) would benefit the users more.

The dropdown with configurable page lengths is definitely a good component on its own, without a doubt, but moving the 100th (or whatever will be the largest number) item to the second page will still be a problem.

**Comment by clarkepaul** [comment]
I have another suggestion outside of fixing the extensions which are similar to @fspringveldt's but instead of a "move to page" have a field in the edit view of an item, something like... 
Display order: 534
It would be a nightmare moving an item which is at 811 to be order number 1 otherwise.

Something else to consider for the UI/design is that this should function for smaller screens, but that should come out in the design.

**Comment by robbieaverill** [comment]
Ok, pull request at #201 up for review. This takes the approach of adding a paginator with dev configurable page sizes.

**Comment by robbieaverill** [comment]
We've proposed to solve this problem for now by using the GridFieldConfigurablePaginator, which was merged in #201 and has a SS4 compatible version being merged up via #208.

Thanks to everybody for their input on this! Hopefully this solution will help in future.
